packages/core/epoch/epoch.go

```diff
@@ -65,10 +65,9 @@ func (i Index) StartTime() time.Time {
 	return time.Unix(startUnix, 0)
 }
 
-// EndTime calculates the end time of the given epoch.
+// EndTime returns the latest possible timestamp for an Epoch. Anything with higher timestamp will belong to the next epoch.
 func (i Index) EndTime() time.Time {
-	endUnix := GenesisTime + int64(i-1)*Duration + Duration - 1
-	return time.Unix(endUnix, 0)
+	return i.StartTime().Add(time.Duration(Duration) * time.Second).Add(-1)
```

Comment:
```suggestion
        endUnix := GenesisTime + int64(i)*Duration
        // we subtract 1 nanosecond from the next epoch to get the latest possible timestamp for epoch i
	return time.Unix(endUnix, 0).Add(-1)
```

This should work, is in line with the style of `StartTime` and be slightly more performant.